aspnetcore/security/authentication/mfa.md

```diff
@@ -0,0 +1,562 @@
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+	services.AddDbContext<ApplicationDbContext>(options =>
+		options.UseSqlite(
+			Configuration.GetConnectionString("DefaultConnection")));
+	
+	services.AddIdentity<IdentityUser, IdentityRole>(
+		options => options.SignIn.RequireConfirmedAccount = false)
+	 .AddEntityFrameworkStores<ApplicationDbContext>()
+	 .AddDefaultTokenProviders();
+
+	services.AddSingleton<IEmailSender, EmailSender>();
+	services.AddScoped<IUserClaimsPrincipalFactory<IdentityUser>, 
+		AdditionalUserClaimsPrincipalFactory>();
+
+	services.AddAuthorization(options =>
+	{
+		options.AddPolicy("TwoFactorEnabled",
+			x => x.RequireClaim("TwoFactorEnabled", "true")
+		);
+	});
+
+	services.AddRazorPages();
+}
+```
+
+The `AdditionalUserClaimsPrincipalFactory` class adds the `TwoFactorEnabled` claim to the user claims after a successful login. This is only added after a login. The value is read from the database. This is added here because the user should only access the higher protected view, if the identity has logged in with MFA. If the database view was read from the database directly instead of using the claim, it would be possible to access the view without MFA directly after activating the MFA.
+
+```csharp
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace IdentityStandaloneMfa
+{
+    public class AdditionalUserClaimsPrincipalFactory : UserClaimsPrincipalFactory<IdentityUser, IdentityRole>
+    {
+        public AdditionalUserClaimsPrincipalFactory( 
+            UserManager<IdentityUser> userManager,
+            RoleManager<IdentityRole> roleManager, 
+            IOptions<IdentityOptions> optionsAccessor) 
+            : base(userManager, roleManager, optionsAccessor)
+        {
+        }
+
+        public async override Task<ClaimsPrincipal> CreateAsync(IdentityUser user)
+        {
+            var principal = await base.CreateAsync(user);
+            var identity = (ClaimsIdentity)principal.Identity;
+
+            var claims = new List<Claim>();
+
+            if (user.TwoFactorEnabled)
+            {
+                claims.Add(new Claim("TwoFactorEnabled", "true"));
+            }
+            else
+            {
+                claims.Add(new Claim("TwoFactorEnabled", "false"));
+            }
+
+            identity.AddClaims(claims);
+            return principal;
+        }
+    }
+}
+```
+
+Because the Identity service setup changed in the `Startup` class, the layouts of the Identity need to be updated. Scaffold the Identity pages into the app. Define the layout in the *Identity/Account/Manage/_Layout.cshtml* file.
+
+```cshtml
+@{
+    Layout = "/Pages/Shared/_Layout.cshtml";
+}
+```
+
+Also add the _layout for all the manage pages from the Identity Pages.
```

Comment:
```suggestion
Also assign the layout for all the manage pages from the Identity pages:
```

Reply:
As this touches templates, @HaoK should also review

Reply:
text fixed